src/datamodel/validator/UUri.cpp

```diff
@@ -10,3 +10,179 @@
+		return true;
+	}
+	if ((uuri.ue_id() & 0xFFFF) == 0xFFFF) {  // service ID
+		return true;
+	}
+	if ((uuri.ue_id() & 0xFFFF0000) == 0) {  // service instance ID
+		return true;
+	}
+	if (uuri.ue_version_major() == 0xFF) {
+		return true;
+	}
+	if (uuri.resource_id() == 0xFFFF) {
+		return true;
+	}
+	return false;
+}
+
+ValidationResult isValid(const v1::UUri& uuri) {
+	{
+		auto [valid, reason] = isValidRpcMethod(uuri);
+		if (valid) {
+			return {true, std::nullopt};
+		}
+	}
+
+	{
+		auto [valid, reason] = isValidRpcResponse(uuri);
+		if (valid) {
+			return {true, std::nullopt};
+		}
+	}
+
+	{
+		auto [valid, reason] = isValidPublishTopic(uuri);
+		if (valid) {
+			return {true, std::nullopt};
+		}
+	}
+
+	return isValidNotification(uuri);
+}
+
+ValidationResult isValidRpcMethod(const v1::UUri& uuri) {
+	if (uuri.authority_name().empty()) {
+		return {false, Reason::EMPTY};
+	}
+
+	// disallow wildcards
+	if (uses_wildcards(uuri)) {
+		return {false, Reason::DISALLOWED_WILDCARD};
+	}
+
+	// check resource ID [0x0001, 0x7FFF]
+	if (uuri.resource_id() == 0 || uuri.resource_id() > 0x7FFF) {
+		return {false, Reason::BAD_RESOURCE_ID};
+	}
+
+	return {true, std::nullopt};
+}
+
+ValidationResult isValidRpcResponse(const v1::UUri& uuri) {
+	if (uuri.authority_name().empty()) {
+		return {false, Reason::EMPTY};
+	}
+
+	// disallow wildcards
+	if (uses_wildcards(uuri)) {
+		return {false, Reason::DISALLOWED_WILDCARD};
+	}
+
+	if (uuri.resource_id() != 0) {
+		return {false, Reason::BAD_RESOURCE_ID};
+	}
+	return {true, std::nullopt};
+}
+
+ValidationResult isValidDefaultSource(const v1::UUri& uuri) {
+	return isValidRpcResponse(uuri);
+}
+
+ValidationResult isValidPublishTopic(const v1::UUri& uuri) {
+	if (uuri.authority_name().empty()) {
+		return {false, Reason::EMPTY};
+	}
+
+	// disallow wildcards
+	if (uses_wildcards(uuri)) {
+		return {false, Reason::DISALLOWED_WILDCARD};
+	}
+
+	if ((uuri.resource_id() & 0xFFFF) < 0x8000) {
+		return {false, Reason::BAD_RESOURCE_ID};
+	}
+
+	return {true, std::nullopt};
+}
+
+ValidationResult isValidNotification(const v1::UUri& uuri) {
+	if (uuri.authority_name().empty()) {
+		return {false, Reason::EMPTY};
+	}
+
+	// disallow wildcards
+	if (uses_wildcards(uuri)) {
+		return {false, Reason::DISALLOWED_WILDCARD};
+	}
+
+	if ((uuri.resource_id() & 0xFFFF) < 0x8000) {
+		return {false, Reason::BAD_RESOURCE_ID};
+	}
+
+	return {true, std::nullopt};
+}
+
+ValidationResult isValidSubscription(const v1::UUri& uuri) {
+	if (uuri.authority_name().empty()) {
+		return {false, Reason::EMPTY};
+	}
+
+	if ((uuri.resource_id() & 0xFFFF) < 0x8000) {
+		return {false, Reason::BAD_RESOURCE_ID};
+	}
+
+	return isValidPublishTopic(uuri);
+}
+
+ValidationResult isEmpty(const v1::UUri& uuri) {
+	if (!std::all_of(uuri.authority_name().begin(), uuri.authority_name().end(),
+	                 isspace)) {
+		return {false, Reason::EMPTY};
+	}
+
+	if (uuri.ue_id() != 0) {
+		return {false, Reason::RESERVED_RESOURCE};
```

Comment:
I just noticed that `RESERVED_RESOURCE` is being used for a non-zero `ue_id`. Incidentally, I think 0 is allowed for both `ue_id` and `resource_id` under some circumstances.

Reply:
TBH this is really minor. I'm not sure if I would stop the review over it.

Reply:
Good, because that check is part of isEmpty() and checking for an empty UUri that has a non-zero ue_id seems right - although we could argue that the error code is wrong.  But that one seems "fitting" unless I wanted to make a new one.

For the record - I'm leaving it unchanged.